ownership first draft

Connortsui20 opened the pull request:
Currently only the first section first draft here

This is following the rust book pretty heavily. It seems not too bad, though we do need to decide if we want to introduce string literals in lecture 1 or here (hw1 doesn't need string literals)

jurplel commented:
> This is following the rust book pretty heavily. It seems not too bad, though we do need to decide if we want to introduce string literals in lecture 1 or here (hw1 doesn't need string literals)

I think we can decide this based on how full this lecture ends up (i.e. according to timing). First lecture is already quite full, but if we have a lot to cover here then we can move strings to lecture 1.

Connortsui20 commented:
Really packing a lot in here.

Topics we need to cover:
- Ownership rules
- Move semantics
- References and Borrowing
- Slices
- String
- Vec basics (for the homework)

jurplel commented:
Notes:
- If they spend over an hour on the assignment (there definitely will be people), and they tell us, what do we do then? What is our end goal?
- "When the owner goes out of scope, the value will be dropped."---this point is kinda unclear from just the slides, it maybe could be improved or an example might be provided.
- The transition to discussion of string literals is unmotivated and confusing. I think it needs to be changed. We haven't really discussed string literals in detail, and it isn't clear why this is a problem at this point for ownership (which is what the slides seem to imply?). Basically, lecture flow issue here.
- We're drawing a lot of distinctions between Strings and literals. I want to make it clear to people while teaching that these are *not the only types of strings in Rust*. I think the tone of the slides should reflect this, especially on introduction. These can be thought of as the introductory forms of strings, maybe? Something to think about
- "To support a mutable piece of text, we need to allocate on the heap"---this isn't quite true, since you can have fixed-length strings on the stack. I think maybe you need to say something more like "resizable" or "fully mutable"
- "This means we must request memory from the allocator at runtime---We need a way of returning the memory when we're done using it"---this segue is really janky. I would suggest going through malloc and free, rust approach to memory, with basic types before introduction this String stuff at all. That would help motivate String's introduction. Introducing malloc/free in the middle is really weird
- If we forget to `free`, we waste memory->If we forget to `free`, we leak memory
- another bullet point before introducing the idea, maybe like "If we always need to free memory we have allocated, maybe we can automate this manual work"
- "What does a "copy" even mean?" I don't think this is needed. It doesn't lead *directly* into the next slide, and most people will have an intuition for what a copy is. Presumably we'd say this out loud.
- In addition, in the same place as that bullet point, I would like something like "Maybe...but this is non-obvious extra performance cost"
- Pointer aliasing slide should be like "One option"---it doesn't say this doesn't happen in Rust, it almost reads like we're explaining what actually happens.
- Pointer aliasing slide should mention this helps us get around the cost of copying all of the memory. That is how we should lead into this slide from the multiple variables same string slide, I think
- Pointer aliasing slide 2: Motivate the double free with "following the drop rules from above" or something.
- How do we solve this problem? Motivation slide before "one owner at a time"
- There is a design choice that is implied by this: Rust will never create a "deep" copy of your data.-> I think this can be better. I prefer "This is an intentional design choice in Rust---there will never be a "deep" copy (of the heap data) without you explicitly asking for it---no hidden costs!"
- "This is a common design philosophy in Rust" maybe add to the slide?
- "This copies _all_ of the data contained in `s1`"->"This copies the data contained in `s1`---heap and stack"
- "There is no difference between a shallow copying and deep copying here, so there is no need to call clone"->I am not sure we should add this bullet point---seems like its missing the main idea. The main idea is that copying an integer isn't a thing---integers fit in registers and are incredbily cheap to copy.
- "Passing a variable to a function will move or copy, just as assignment does." Maybe uncomment the println and show that it won't compile
- I think you should split the gives ownership and takes and give back examples into two slides. It is a lot to dig through in one slide---reminds me of 213 slides D:
- Highlight we're really only going to have *move*, *clone*, and *copy* operations at the end of this section as a conclusion
- Moving is somewhat tedious slide---mention that we don't want to clone for performance reasons.
- references slide 1 needs a redo. Both bullet points are not worded well (sorry). Should motivate why clone, copy, move won't cover all cases here. Maybe you can move some of the move motivation from the last section. Then say, okay, so we can *share* memory using references.
- "We do not own `s1` if we just have `&s1`"-> this slide is referring to behavior from calculate_length. Who is we? "This means `s1` will not be dropped when we stop using `&s1`" again who is we? This I think refers to the helper.
- This next slide is empty. I would suggest we combine them, drop the slide header for more room, and optimize it to a few denser bullet points.
- "references are just pointers"->"references are just like pointers"
- "Sometimes people will refer to mutable references as exclusive references"-> I suggest we drop this bullet--i don't hear anyone call them this and i would not like students to call them this
- "This restriction allows for controlled mutation"-> this isn't specific enough. I suggest we try motivating mutable references being exclusive before saying that they are exclusive. Can we come up with an example why it might be bad?
- "most languages let you mutate whenever you want"->be more specific! With pointers? with references? I'm not sure, this point just seems like it can have more meaning.
- "Rust instead prevents data races at compile time!" perfect
- Great examples
- "new reference"->"new mut reference" in example comment
- It would be cool for the error slides if we could have an animation circling the important bit. Low priority, not sure if its possible with marp easily
- "it will not dangling references"->"it will not allow dangling references"
- "Think many readers of a book, but a single writer (read-write lock)"->"A book being read by multiple people is fine, but if more than one people write they may overwrite each other's work." Read-write lock as a separate sub-bulletpoint, we have room.
- holy crap we're covering slices too. This is a lot. I see why you said to cover examples after class. I completely agree that's how this will end up.
- "we can properly understand string literals"->"we can properly explain how string literals are refered to"??
- "and so we can't mutate them"->awkward wording
- "...type, so we can..."->maybe try to cut use of the word "so" a bit
- "We HIGHLY recommend reading the Rust Book chapter on Ownership"->I think this wording is a little too strong. I don't expect stuco students to do reading outside of class. I would refer to it more as a secondary resource for a better review, recommended reading. I would lay off the highly part.
- "understating"->"understanding"
- Consider downloading a VS Code spell checker (no offense i do it too)
- "Ownership"->"ownership" randomly capitalized on this slide
- "Thanks for showing up!"->Attendence is required. You can say "Thanks for coming!"--less sarcastic, more professional.
- I think moving string literals to week1 is good.
- I would go a step further---I think we should introduce 'String' to them in week 1. I don't think we should say much beyond how it works at a basic level. Introducing both of those will help reduce the amount of new things in this lecture a noticable amount.
- Vec could also be pushed to next week. I think we'll be rushing through it at this rate. I understand we need it for the homework though :(

jurplel reviewed (commented):
^

Connortsui20 commented:
>  If they spend over an hour on the assignment (there definitely will be people), and they tell us, what do we do then? What is our end goal?

Then we will help them a bit more. If they take more than an hour, we will ask them on what things they took longer on. It might be on us that we made it too hard, and in that case we can maybe remove a few puzzles. It might also be because they don't fully understand something, and in that case, we definitely want to make sure they do understand. I think that the first few weeks learning Rust are _so_ important: if we need to go out of our way to give a bit of extra help in the first 2 weeks I'm okay with that.

> "When the owner goes out of scope, the value will be dropped."---this point is kinda unclear from just the slides, it maybe could be improved or an example might be provided.

The point here is to just introduce the words, mainly the fact that Ownership is a set of rules. Then we go through the what and why of those rules, and then finally circle back.

For the string stuff, instead of making it review I just added it as a new topic now. Everyone should know what string literals are from C and 122 and even more from 213, and they should also understand that we cannot mutate them. As for the many types of Rust strings, that is definitely something we mention in lecture and not in the slides, we don't want to overwhelm people. 

> this segue is really janky.

I actually think this can be good. I added a "What 2 `C` functions does this remind you of?" in there to help a bit. That will naturally lead people to think of malloc and free (which all of them should know well), which is then the next slide. The whole reason why Rust has its ownership model is because of the malloc<->free problem. I've added an extra slide to motivate the scopes, let me know if you think this is better.

Copy and aliasing stuff: I added a bunch of stuff in the speaker notes, we can just say it (mainly the performance stuff)

> "There is no difference between a shallow copying and deep copying here, so there is no need to call clone"->I am not sure we should add this bullet point---seems like its missing the main idea. The main idea is that copying an integer isn't a thing---integers fit in registers and are incredbily cheap to copy.

I sort of agree? It's a bit wordy and it's not super important. But the logical jump from integers being cheap to copy to why integers are allowed to do this and not String might not be _immediately_ obvious. The point here is to motivate why String is not allowed to do this: everyone sort of understands integer copying already from C. Not sure how else we should motivate this though.

> Highlight we're really only going to have move, clone, and copy operations at the end of this section as a conclusion

Not sure I understand this

> Moving is somewhat tedious slide---mention that we don't want to clone for performance reasons.

Also not sure why this is relevant to this slide, though after the changes we will have said that enough in the past slides.

Also don't apologize when my wording is awful lmao, I completely missed that

> This next slide is empty. I would suggest we combine them, drop the slide header for more room, and optimize it to a few denser bullet points.

Did some refactoring, but kept it as 2 slides

> "Sometimes people will refer to mutable references as exclusive references"-> I suggest we drop this bullet--i don't hear anyone call them this and i would not like students to call them this

This is actually a common term, and I remember seeing [discussion](https://docs.rs/dtolnay/0.0.11/dtolnay/macro._02__reference_types.html) about changing the wording to "shared" and "exclusive" references instead of "immutable" and "mutable" references.  I'll take it out of the slides, but it is something we should say in lecture.

> "This restriction allows for controlled mutation"-> this isn't specific enough. I suggest we try motivating mutable references being exclusive before saying that they are exclusive. Can we come up with an example why it might be bad?

I changed it to "explicit mutation". I think the previous slides do motivate this? Any other examples have to be explained with concurrency, which we definitely do not want to put in the slides. We can just say it in lecture.

> "most languages let you mutate whenever you want"->be more specific! With pointers? with references? I'm not sure, this point just seems like it can have more meaning.

Changed it to "mutate anything," and we can say that this is values, pointers, variables, etc.

>  It would be cool for the error slides if we could have an animation circling the important bit. Low priority, not sure if its possible with marp easily

Yeah I don't think that is possible, but its not like we can't walk up to the slides and point directly (which is what I was planning to do for every single error message slide)

>  "We HIGHLY recommend reading the Rust Book chapter on Ownership"->I think this wording is a little too strong. I don't expect stuco students to do reading outside of class. I would refer to it more as a secondary resource for a better review, recommended reading. I would lay off the highly part.

Students can choose not to read source material, like every other class. This week especially, since the homework is going to be light, we should encourage them to get more familiar with the concept of ownership by reading. Students who don't care will not read, and that is fine. But students who want to get a better understand will see this and maybe get the extra push to read the source content.

I actually now think that it is possible to do all of this in 50 minutes. See `plan.md`. This is now relying on information I shouldn't be relying on - but given that we have all masters and seniors in our course, they should understand what string literals are. Also, they should be familiar with dynamic data types stored on the heap (from Python or C++). On top of this, they should all be familiar with read write locks. This is actually why I recommend reading the book again: there is only so much we can explain about ownership in one day. You really need to get a feeling over a span of time for ownership, so it's fine if students leave this lecture confused about a few things (mainly the whys, not the whats). If we somehow figured out a way to teach properly teach ownership in 50 minutes we could publish a book.

So basically what I'm saying is that I think it's fine for us to be a bit speedy in this lecture - that's why I planned a ownership review in the next lecture.

Thanks for the review!

jurplel commented:
> Students can choose not to read source material, like every other class. This week especially, since the homework is going to be light, we should encourage them to get more familiar with the concept of ownership by reading. Students who don't care will not read, and that is fine. But students who want to get a better understand will see this and maybe get the extra push to read the source content.

Okay, sure. Maybe we can leave highly as a lowercase word? It might be a little unprofessional—just a feeling. We can stress it in lecture, though!

> I changed it to "explicit mutation". I think the previous slides do motivate this? Any other examples have to be explained with concurrency, which we definitely do not want to put in the slides. We can just say it in lecture.

"This restriction allows for explicit mutation" I am not sure exactly what you meant by this line. What I meant by my comment is more that it would be better to write out exactly what "controlled mutation" means/what it does for us. I am not sure—maybe let's chat about this one. I just am not sure what I would do with this bullet point besides read it—I am not sure what it means. I feel that ideally it should be obvious what it means when reading the slides.



> Highlight we're really only going to have move, clone, and copy operations at the end of this section as a conclusion
Not sure I understand this

I am suggesting a slide that concludes the section before we introduce borrowing, saying "Okay, we can move, clone and copy". Cloning is expensive, copying only make sense with primitive types, and moves are inconvenient/impractical in a lot of circumstances, cue references.

> I actually think this can be good. I added a "What 2 C functions does this remind you of?" in there to help a bit. That will naturally lead people to think of malloc and free (which all of them should know well), which is then the next slide. The whole reason why Rust has its ownership model is because of the malloc<->free problem. I've added an extra slide to motivate the scopes, let me know if you think this is better.

I think it's better!

> that is definitely something we mention in lecture and not in the slides, we don't want to overwhelm people.

I think that is sort of a dangerous mindset—when people want to review the content, they will look back at the lecture slides. We should try to make sure all of the important stuff is in the slides. That being said, I think the current situation with strings is okay.

Minor correction: "Drop function -> drop method"

As for my finishing thoughts on moving things to different lectures, I actually think it is okay. They will know what Strings are pretty quickly. We'll have to quickly reintroduce them in this lecture anyway, which will probably take a similar amount of time. Better to have extra time in week 1 for logistic issues and such.

Nice, I think it's about ready to ship. Would like drudo to go through it.

Connortsui20 commented:
Adding a recap slide for the ownership section.

> I think that is sort of a dangerous mindset—when people want to review the content, they will look back at the lecture slides. We should try to make sure all of the important stuff is in the slides. That being said, I think the current situation with strings is okay.

I agree, but it's not like people are going to come to this lecture to review the 10+ types of strings in Rust. They're also never going to use more than &str and String anyways... I think it would be a bad idea to even mention stuff like Cow<str> and OsStr. If they want to look at documentation, they always can.

> Minor correction: "Drop function -> drop method"

You actually cannot call drop as a method, its one of the special traits
